Add a `max_dust_htlc_exposure_msat` (#919)
lightning/bolts: 02-peer-protocol.md

  transaction, AND is past this fulfillment deadline:
    - MUST fail the channel.

### Bounding exposure to trimmed in-flight HTLCs: `max_dust_htlc_exposure_msat`

> When a HTLC is present in a channel is below the "trimmed" threshold in [BOLT3 #3](03-transactions.md),

Reviewer: nit:
```suggestion
When an HTLC in a channel is below the "trimmed" threshold in [BOLT3 #3](03-transactions.md),
```

the HTLC is not claimable on-chain, instead being turned into additional miner
fees if either party unilaterally closes the channel. Because the threshold is
per-HTLC, the total exposure to such HTLCs may be substantial if there are many
dust HTLCs present when the channel is force-closed.

This can be exploited in griefing attacks or even in miner-extractable-value attacks,
if the malicious entity avails <sup>[mining capabilities](https://lists.linuxfoundation.org/pipermail/lightning-dev/2020-May/002714.html)</sup>.

The total exposure is given by the following back-of-the-envelope computation:

	counterparty's `max_accepted_htlcs` * (`HTLC-success-kiloweight` * opener's `feerate_per_kw` + counterparty's `dust_limit_satoshis`)
		+ holder's `max_accepted_htlcs` * (`HTLC-timeout-kiloweight` * opener's `feerate_per_kw` + counterparty's `dust_limit_satoshis`)

> To mitigate this scenario, a `max_dust_htlc_exposure_msat` can be apply at

Reviewer: nit:
```suggestion
To mitigate this scenario, a `max_dust_htlc_exposure_msat` can be applied at
```

HTLC sending, forwarding and receiving.

A node:
  - upon an incoming HTLC:

>     - if a HTLC's `amount_msat` is inferior to the counterparty's `dust_limit_satoshis` plus the HTLC-timeout fee at the `feerate_per_kw`:

Reviewer: nit: if "the" HTLC's (...)

And same for the other requirements below. Technically we should probably say "a hash-time-lock contract" but the spec uses "an HTLC' everywhere, so we should be consistent.

### Bounding exposure to trimmed in-flight HTLCs: `max_dust_htlc_exposure_msat`

When a HTLC is present in a channel is below the "trimmed" threshold in [BOLT3 #3](03-transactions.md),
the HTLC is not claimable on-chain, instead being turned into additional miner
fees if either party unilaterally closes the channel. Because the threshold is
per-HTLC, the total exposure to such HTLCs may be substantial if there are many
dust HTLCs present when the channel is force-closed.

This can be exploited in griefing attacks or even in miner-extractable-value attacks,
if the malicious entity avails <sup>[mining capabilities](https://lists.linuxfoundation.org/pipermail/lightning-dev/2020-May/002714.html)</sup>.

The total exposure is given by the following back-of-the-envelope computation:

	counterparty's `max_accepted_htlcs` * (`HTLC-success-kiloweight` * opener's `feerate_per_kw` + counterparty's `dust_limit_satoshis`)
		+ holder's `max_accepted_htlcs` * (`HTLC-timeout-kiloweight` * opener's `feerate_per_kw` + counterparty's `dust_limit_satoshis`)

To mitigate this scenario, a `max_dust_htlc_exposure_msat` must be apply at
HTLC sending, forwarding and receiving.

> A node:

Reviewer 2: imo this whole section should be moved into the `update_add_htlc` requirements, as they're actions that are triggered by the receipt of this message.

Author: Well, in fact it's not exactly clear that it should be in the `update_add_htlc` requirements as it's a relay requirement, to me in the same category of `cltv_expiry_delta` selection. At least was my thinking when I wrote those lines. 

Though it can be a source of confusion and maybe we should even add a new BOLT4 message. Due to lack of that, I remember it was hard to assess that `max_dust_htlc_exposure_msat` was well in-place when I verified fixes against Eclair in a black-box way.

Reviewer 2: yes but you don't enact these spec requirements until a `update_add_htlc` message is received.

	- SHOULD fail this HTLC once it's committed
	- SHOULD NOT reveal a preimage for this HTLC
    - if a HTLC's `amount_msat` is inferior to the holder's `dust_limit_satoshis` plus the HTLC-success fee at the `dust_buffer_feerate`:
      - if the `amount_msat` plus the `dust_balance_on_holder_tx` is superior to `max_dust_htlc_exposure_msat`:
	- SHOULD fail this HTLC once it's committed
        - SHOULD NOT reveal a preimage for this HTLC
  - upon an outgoing HTLC:
    - if a HTLC's `amount_msat` is inferior the counterparty's `dust_limit_satoshis` plus the HTLC-success fee at the `dust_buffer_feerate`:
      - if the `amount_msat` plus the `dust_balance_on_counterparty_tx` is superior to `max_dust_htlc_exposure_msat`:
        - SHOULD NOT send this HTLC
	- SHOULD fail this HTLC if it's forwarded
    - if a HTLC's `amount_msat` is inferior to the holder's `dust_limit_satoshis` plus the HTLC-timeout fee at the `dust_buffer_feerate`:
      - if the `amount_msat` plus the `dust_balance_on_holder_tx` is superior to `max_dust_htlc_exposure_msat`:
        - SHOULD NOT send this HTLC
	- SHOULD fail this HTLC if it's forwarded

> `dust_buffer_feerate` is defined as the maximum of either 2530 sats per kWU or

Reviewer 3: I don't think the spec needs to be this prescriptive. This is what we do in LDK, but others can/should do other things. I think we the spec should just talk generally about `feerate_per_kw` and note that in calculating you SHOULD provide some buffer for future feerate increases.

For channels that don't use `option_anchors_zero_fee_htlc_tx`, an increase of
the `feerate_per_kw` may trim multiple htlcs from commitment transactions,
which could create a large increase in dust exposure. If `option_anchors_zero_fee_htlc_tx`
has been negotiated, the HTLC feerate is implied to be 0, the checks should

> still be applied as a node can be exposed due to the sum of the trimmed HTLCs.

Reviewer: This sentence is quite unclear to me: which checks are you talking about? I don't think this new sentence is necessary at all to be honest, the requirements above already apply to all commitment types, and this paragraph only says that without `option_anchors_zero_fee_htlc_tx` you should be extra cautious when the feerate changes.

Author: Removed this sentence on latest commit `7705356`, if we think it doesn't add clarification.

The node _not responsible_ for paying the Bitcoin fee:
  - MUST NOT send `update_fee`.

A sending node:
  - if the `dust_balance_on_counterparty_tx` at the new `dust_buffer_feerate` is superior to `max_dust_htlc_exposure_msat`:
    - MAY NOT send `update_fee`
    - MAY fail the channel

>   - if the `dust_balance_on_holder_tx` at the new `dust_buffer_feerate` is superior to the `max_dust_htlc_exposure_msat`:

Reviewer 2: This should be 
```suggestion
  - if the `dust_balance_on_holder_tx` at the new `feerate_per_kw` is superior to the `max_dust_htlc_exposure_msat`:
```

The real question is if the new feerate pushes the balance over the exposure limit, not if the new feerate + buffer does. Pushing a feerate update where the new buffer ceiling is above the current should just fail to add any new dusty htlcs, but will leave the existing ones without a problem. Otherwise what was the point of having the buffer in the first place?

Author: Yes your right that's the spec doesn't match our LDK code on that: https://github.com/rust-bitcoin/rust-lightning/blob/7aa2caccd884cd7d40ee146323c2a65b7ea39407/lightning/src/ln/channel.rs#L3130

> Pushing a feerate update where the new buffer ceiling is above the current should just fail to add any new dusty htlcs, but will leave the existing ones without a problem. 

Yes that's right. If the inbound/outbound feerate is above the the dust buffer feerate ceiling, the dust exposure on counterparty and holder commitment transactions should be evaluated. If one of them is above the dust HTLC exposure the channel _could_  be fail (I think a MAY is better).

We had multiple back-and-forth on the LDK-side on the design of `update_fee` reception mitigations.

Updating the PR.

Author: Updated at 6a1cb21. If you think there are more corrections needed, let me know.

  current commitment transaction:
    - SHOULD fail the channel,
      - but MAY delay this check until the `update_fee` is committed.
  - if the `update_fee` is above the `feerate_per_kw`:
    - if the dust balance on counterparty transaction at the current `feerate_per_kw` is superior to `max_dust_htlc_exposure_msat`:
      - MAY fail the channel
    - if the dust balance on holder transaction at the current `feerate_per_kw` is superior to `max_dust_htlc_exposure_msat`:
      - MAY fail the channel

For pre-`option_anchors_zero_fee_htlc` channels, the second-stage HTLC transaction
fee is function of the negotiated `feerate_per_kw`. To avoid a node being
significantly more exposed to trimmed balance in case of feerate increases when
several HTLCs are pending which are near the dust limit, it is recommended to

> calculate that `feerate_per_kw` with some buffer.

Reviewer: nit: I'd remove this duplicated paragraph (already there above).

  current commitment transaction:
    - SHOULD fail the channel,
      - but MAY delay this check until the `update_fee` is committed.
  - if the `dust_balance_on_counterparty_tx` at the new `dust_buffer_feerate` is superior to `max_dust_htlc_exposure_msat`:

>     - MAY fail the channel

Reviewer 2: ```suggestion
    - SHOULD fail the channel
```

Author: See the rational just under. That's a MAY as automatic-closure might bring newer issues. A third-party to a channel could relay and withhold HTLCs across the link to provoke a close.

See also Eclair new config settings `close-on-update-fee-dust-exposure-overflow` : https://github.com/ACINQ/eclair/pull/1985/files, which is false by default. Though in LDK we do auto-close by default.

As it's arbitrating between two safety risks, I think it's wiser to defer the choice to the implementator.

Reviewer 2: Well 'SHOULD' is still a deferral to the implementor, just a more strongly worded form!

  - if the `dust_balance_on_holder_tx` at the new `dust_buffer_feerate` is superior to the `max_dust_htlc_exposure_msat`:

>     - MAY fail the channel

Reviewer 2: ```suggestion
    - SHOULD fail the channel
```

Author: See https://github.com/lightningnetwork/lightning-rfc/pull/919#discussion_r721767658
